// pgloader/dist/resources.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.updateHtml = exports.updateLinks = exports.fetchResources = exports.getLinks = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _url = require('url');

var _url2 = _interopRequireDefault(_url);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _cheerio = require('cheerio');

var _cheerio2 = _interopRequireDefault(_cheerio);

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _common = require('./common');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

/*
eslint no-shadow: ["error", { "allow": ["data", "url", "fileName", "error"] }]
*/
/*
eslint-env es6
*/

var log = (0, _debug2.default)('page-loader:');

var getLinks = function getLinks(html, selector, predicate) {
  var $ = _cheerio2.default.load(html);
  return $(selector).toArray().map(function (el) {
    var src = el.attribs.src || el.attribs.href;
    return { type: el.name, src: src };
  }).filter(function (el) {
    return !predicate.test(el.src);
  });
};

var updateLinks = function updateLinks(linksColl, dirName) {
  return linksColl.reduce(function (acc, el) {
    var _nodePath$parse = _path2.default.parse(el.src),
        dir = _nodePath$parse.dir,
        ext = _nodePath$parse.ext,
        name = _nodePath$parse.name;

    var newEl = _extends({}, el);
    var newName = '' + _path2.default.join(dir, name).replace(/\W/g, '-').slice(1) + ext;
    newEl.localSrc = _path2.default.format({
      root: '/ignored',
      dir: dirName,
      base: newName,
      ext: 'ignored'
    });
    newEl.fileName = newName;
    return [].concat(_toConsumableArray(acc), [newEl]);
  }, []);
};

var updateHtml = function updateHtml(html, linksColl) {
  var result = linksColl.reduce(function (acc, el) {
    var attrib = el.type === 'link' ? 'href' : 'src';
    var $ = _cheerio2.default.load(acc);
    $('[' + attrib + '="' + el.src + '"]').removeAttr(attrib).attr(attrib, el.localSrc);
    return $.html();
  }, html);

  return result + '\n';
};

var fetchResources = function fetchResources(html, url, resourcesDir, htmlFileName) {
  var linksColl = getLinks(html, 'img[src], script[src], link[href]', /^(\w+:)?\/{2,}/);
  var updatedLinksColl = updateLinks(linksColl, resourcesDir);
  var updatedHtml = updateHtml(html, updatedLinksColl);
  var coll = [{ type: 'html', data: Promise.resolve(updatedHtml), location: htmlFileName }];
  var result = updatedLinksColl.reduce(function (acc, el) {
    var responseType = el.type === 'img' ? 'stream' : 'json';

    var _nodeUrl$parse = _url2.default.parse(url),
        protocol = _nodeUrl$parse.protocol,
        host = _nodeUrl$parse.host;

    log('Fetching... ' + protocol + '//' + host + el.src);
    var newEl = {
      type: el.type,
      data: (0, _common.getResponse)(protocol + '//' + host + el.src, responseType),
      location: '' + resourcesDir + _path2.default.sep + el.fileName
    };
    return [].concat(_toConsumableArray(acc), [newEl]);
  }, coll);

  return result;
};

exports.getLinks = getLinks;
exports.fetchResources = fetchResources;
exports.updateLinks = updateLinks;
exports.updateHtml = updateHtml;
